### PR TITLE
[coordinator] Make <500 endpoint errors as invalid params error

### PR DESCRIPTION
**Pull request by Antanukas**
**What this PR does / why we need it**: 

map <500 response codes to 400.

**Special notes for your reviewer**:

Makes sure 4xx errors are wrapped in InvalidParamsError that eventually gets mapped to 400 HTTP status code.
This is important because prometheus [scraper retries 5xx errors](https://github.com/prometheus/prometheus/blob/fac1b573349ac4d35c011a891e2bb58f560f9d84/storage/remote/client.go#L241). Before this change any errors returned from endpoint would have been mapped to 500.

Storage layer abstracts away HTTP so it's not possible to just respond with same status codes as in response from the endpoint. However it seems enough to map to 400 or 500 based on Prometheus scraper implementation.

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
None
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
None
```

### CODE REVIEW

File: src/query/storage/promremote/storage.go [hunk 1]
```diff
@@ -164,8 +164,14 @@ func (p *promStorage) writeSingle(
...
 		}
-		return fmt.Errorf("expected status code 2XX: actual=%v, address=%v, resp=%s",
-			resp.StatusCode, address, response)
+		genericError := fmt.Errorf(
+			"expected status code 2XX: actual=%v, address=%v, resp=%s",
+			resp.StatusCode, address, response,
+		)
+		if resp.StatusCode < 500 && resp.StatusCode != http.StatusTooManyRequests {
+			return xerrors.NewInvalidParamsError(genericError)
+		}
```
Comment (vpranckaitis) [inline src/query/storage/promremote/storage.go:173]: There actually is an error type which carries the response code. Though I'm not exactly sure if somewhere else in the stack it won't get wrapped in some other error and as such lose the status code.

https://github.com/m3db/m3/blob/bd4237c36cfaca6cdf6620088b1eadbc885274b9/src/x/net/http/errors.go#L60-L62

Also, should probably discuss with others if this error should mostly be used API/http layers of the codebase, or if it's ok to use it deeper.

File: src/query/storage/promremote/storage_test.go [hunk 2]
```diff
@@ -248,6 +249,28 @@ func TestWriteBasedOnRetention(t *testing.T) {
...
 		assert.NotNil(t, promLongRetention2.GetLastWriteRequest())
 	})
+
+	t.Run("wrap non 5xx errors as invalid params error", func(t *testing.T) {
+		reset()
+		promLongRetention.SetError("test err", http.StatusForbidden)
+		err := sendWrite(storagemetadata.Attributes{
+			Resolution: 10 * time.Minute,
+			Retention:  8760 * time.Hour,
+		})
```
Comment (vpranckaitis) [inline src/query/storage/promremote/storage_test.go:259]: nit: There are quite a few places in the tests where `storagemetadata.Attributes` is defined inline. I think this adds quite a bit of noise. Maybe assign a few `storagemetadata.Attributes` variations to to variables and use those instead? E.g. 

```suggestion
		promLongRetention.SetError("test err", http.StatusForbidden)
		err := sendWrite(longRetentionStorageMetadata)
```

File: src/query/storage/promremote/storage_test.go [hunk 3]
```diff
@@ -248,6 +249,28 @@ func TestWriteBasedOnRetention(t *testing.T) {
...
+		require.Error(t, err)
+		assert.True(t, xerrors.IsInvalidParams(err))
+	})
+
+	t.Run("429 should not be wrapped as invalid params", func(t *testing.T) {
```
Comment (vpranckaitis) [inline src/query/storage/promremote/storage_test.go:264]: The test function is called `TestWriteBasedOnRetention`, but the two new tests actually look into error responses/status codes. I think they should be in a separate test function, e.g. `TestStatusCodeHandling` or similar.

File: src/query/storage/promremote/storage.go (hunk 1 above)
Comment (Antanukas) [reply src/query/storage/promremote/storage.go:173]: I don't think it's a good idea to throw it from storage layer. It seems currently its only thrown form handlers and middleware which are HTTP related things.

Comment (Antanukas) [reply src/query/storage/promremote/storage.go:173]: Also current error handling in HTTP layer is unique for this handler https://github.com/m3db/m3/blob/013972931d01f3019a7089680e8fa2d135bc9424/src/query/api/v1/handler/prometheus/remote/write.go#L321

So I just played it safe to not affect other storages.

**Review by vpranckaitis — APPROVED** [review]

**Pull request merged**